internal/provider/host_resource.go

```diff
@@ -0,0 +1,390 @@
+
+	err = data.ParseHttpResponse(body)
+	if err != nil {
+		diags.AddError("error while parsing the json response: ", err.Error())
+		return diags
+	}
+
+	return diags
+}
+
+func (r HostResource) AssociateGroup(data HostResourceModelInterface) diag.Diagnostics {
+	req_data, diags := MakeReqData(data)
+	resp, _, err := r.client.doRequest(http.MethodPost, data.GetURL()+"/groups/", req_data)
+	diags.Append(IsResponseValid(resp, err, http.StatusNoContent)...)
+
+	return diags
+}
+
+func (r HostResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data HostResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(r.CreateHost(&data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+
+	if IsValueProvided((&data).GroupId) {
+		resp.Diagnostics.Append(r.AssociateGroup(&data)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		// Save updated data into Terraform state
+		resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+	}
+}
+
+func (r HostResource) DeleteHost(data HostResourceModelInterface) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	resp, _, err := r.client.doRequest(http.MethodDelete, data.GetURL(), nil)
+	diags.Append(IsResponseValid(resp, err, http.StatusNoContent)...)
+
+	return diags
+}
+
+func (r HostResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data HostResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	resp.Diagnostics.Append(r.DeleteHost(&data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r HostResource) UpdateHost(data HostResourceModelInterface) diag.Diagnostics {
+	req_data, diags := MakeReqData(data)
+	resp, body, err := r.client.doRequest(http.MethodPut, data.GetURL(), req_data)
+	diags.Append(IsResponseValid(resp, err, http.StatusOK)...)
+
+	err = data.ParseHttpResponse(body)
+	if err != nil {
+		diags.AddError("error while parsing the json response: ", err.Error())
+		return diags
+	}
+
+	return diags
+}
+
+func (r HostResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data HostResourceModel
+	var data_with_URL HostResourceModel
+
+	// Read Terraform plan and state data into the model
+	// The URL is generated once the host is created.
+	// To update the correct host, we retrieve the state data
+	// and append the URL from the state data to the plan data.
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &data_with_URL)...)
+	data.URL = data_with_URL.URL
```

Comment:
This is not needed. The URL will be retrieved from the state and populated in the plan.